aidial_assistant/chain/history.py

```diff
@@ -91,3 +123,63 @@ def to_best_effort_messages(
         )
 
         return messages
+
+    async def trim(
+        self, max_prompt_tokens: int, model_client: ModelClient
+    ) -> "History":
+        extra_results_callback = ModelExtraResultsCallback()
+        stream = model_client.agenerate(
+            self.to_protocol_messages(),
+            extra_results_callback,
+            max_prompt_tokens=max_prompt_tokens,
+            max_tokens=1,
+        )
+        try:
+            async for _ in stream:
+                pass
+        except ReasonLengthException:
+            # Expected for max_tokens=1
+            pass
+
+        if extra_results_callback.discarded_messages:
+            return History(
+                assistant_system_message_template=self.assistant_system_message_template,
+                best_effort_template=self.best_effort_template,
+                scoped_messages=self._skip_messages(
+                    extra_results_callback.discarded_messages
+                ),
+            )
+
+        return self
+
+    def _skip_messages(self, message_count: int) -> list[ScopedMessage]:
+        messages = []
```

Comment:
Add type hints for generic types. Otherwise the type variable will be assumed to be `Unknown` and the typechecker won't be able to hightlight certain typing errors.

Reply:
Added.